orc8r/cloud/go/services/certifier/servicers/certifier.go

```diff
@@ -712,51 +714,95 @@ func (srv *CertifierServer) getPolicyDecisionFromTokenMany(tokens *certprotos.To
-		}
 		resourceEffect := checkResource(reqResource, policyResource)
-		if resourceEffect == certprotos.Effect_DENY {
-			return certprotos.Effect_DENY, nil
-		}
 		// The effects from both action and resource should match
 		// for the effect of the policy to apply to the request.
 		if actionEffect != certprotos.Effect_UNKNOWN &&
 			resourceEffect != certprotos.Effect_UNKNOWN &&
 			actionEffect == resourceEffect {
 			effect = actionEffect
 		}
+
+		if actionEffect == certprotos.Effect_DENY &&
+			resourceEffect == certprotos.Effect_DENY {
+			return certprotos.Effect_DENY, nil
+		}
 	}
 
 	return effect, nil
 }
 
-// checkResource checks if the requested resource is authorized by the policy
-func checkResource(reqResource, policyResource *certprotos.Resource) certprotos.Effect {
-	// The policy does not handle this case, so return UNKNOWN.
-	if reqResource.ResourceType != policyResource.ResourceType {
-		return certprotos.Effect_UNKNOWN
+func getTenantPolicyNetworkResources(ctx context.Context, policy *certprotos.Policy) ([]*certprotos.PolicyResource, error) {
+	var networkResources []*certprotos.PolicyResource
+	for _, policyResource := range policy.Resources.Resources {
+		if t := policyResource.GetTenant(); t != nil {
+			var networks []string
+			for _, i := range t.Tenants {
+				tenant, err := tenants.GetTenant(ctx, i)
+				if err != nil {
+					return nil, status.Errorf(codes.Internal, "%v", err)
+				}
+				networks = append(networks, tenant.Networks...)
+			}
+
+			resource := &certprotos.PolicyResource{
+				Effect:   policyResource.Effect,
+				Action:   policyResource.Action,
+				Resource: &certprotos.PolicyResource_Network{Network: &certprotos.NetworkResource{Networks: networks}},
+			}
+			networkResources = append(networkResources, resource)
+		}
 	}
-	reqPath := reqResource.Resource
-	policyPath := policyResource.Resource
-	if ok, _ := doublestar.Match(policyPath, reqPath); ok {
-		return policyResource.Effect
+	return networkResources, nil
+}
+
+func getRequestResourceType(reqResource *certprotos.RequestResource) constants.ResourceType {
+	reqResourceType := constants.Path
+	switch reqResource.ResourceId.(type) {
+	case *certprotos.RequestResource_NetworkId:
+		reqResourceType = constants.NetworkID
+	case *certprotos.RequestResource_TenantId:
+		reqResourceType = constants.TenantID
 	}
-	return certprotos.Effect_UNKNOWN
+	return reqResourceType
+}
+
+func getPolicyResourceType(policyResource *certprotos.PolicyResource) constants.ResourceType {
+	policyResourceType := constants.Path
+	switch policyResource.Resource.(type) {
+	case *certprotos.PolicyResource_Network:
+		policyResourceType = constants.NetworkID
+	case *certprotos.PolicyResource_Tenant:
+		policyResourceType = constants.TenantID
+	}
+	return policyResourceType
+}
+
+// checkResourceType checks if the policy type applies to the requested type
+func checkResourceType(reqResource *certprotos.RequestResource, policyResource *certprotos.PolicyResource) bool {
```

Comment:
For all these "check x", it's unclear what the true/false mean. Consider renaming them to be something like `checkResourceTypesMatch`, `doResourceTypesMatch`, or even `validateResourceType`

For all cases, the docstring should have something to the effect of "returns true if x, else false"